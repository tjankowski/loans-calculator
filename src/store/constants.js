export const FormFields = {
  amount: "amount",
  duration: "duration",
};

export const FormFieldsValueRange = {
  [FormFields.amount]: [1000, 5000],
  [FormFields.duration]: [6, 36],
};

export const Currency = {
  GBP: {
    key: "GBP",
    label: "Â£",
  },
};
