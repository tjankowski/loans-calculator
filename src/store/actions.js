import quote from "../api";
import { FormFieldsValueRange, FormFields } from "../store/constants";
import debounce from "lodash.debounce";

export const actionTypes = {
  QUOTE_FETCH: "quote.fetch",
  QUOTE_SUCCESS: "quote.success",
  QUOTE_ERROR: "quote.error",
  FORM_INPUT: "form.input",
};

export const useActions = (state, dispatch) => ({
  quote: (amount, duration) => {
    return fetchQuote(dispatch, { amount, duration });
  },
  onFormChange: (field, input) => {
    const value = validateFormInput(field, input);
    dispatch(action(actionTypes.FORM_INPUT, { field, value }));
    if (!value.isError) {
      return fetchQuote(dispatch, {
        amount: state.form[FormFields.amount].value,
        duration: state.form[FormFields.duration].value,
        [field]: value.value,
      });
    }
    return null;
  },
});

const fetchQuote = debounce((dispatch, { amount, duration }) => {
  dispatch(action(actionTypes.QUOTE_FETCH, { amount, duration }));
  return quote(amount, duration)
    .then((data) => {
      dispatch(action(actionTypes.QUOTE_SUCCESS, data));
    })
    .catch((error) => {
      dispatch(action(actionTypes.QUOTE_ERROR, error));
    });
}, 200);

function validateFormInput(field, value) {
  const [min, max] = FormFieldsValueRange[field];
  return {
    isError: value < min || value > max,
    value: Number(value > max ? max : value),
  };
}

export function action(type, payload) {
  return {
    type,
    payload,
  };
}
