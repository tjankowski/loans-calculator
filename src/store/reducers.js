import { actionTypes } from "./actions";
import { FormFields, Currency } from "./constants";

export const initialState = {
  quote: {
    isFetching: false,
    isError: false,
    data: null,
  },
  currency: Currency.GBP,
  form: {
    [FormFields.amount]: {
      value: 2000,
      isError: false,
    },
    [FormFields.duration]: {
      value: 24,
      isError: false,
    },
  },
};

export function reducer(state = initialState, action = { type: null }) {
  switch (action?.type) {
    case actionTypes.FORM_INPUT:
      return {
        ...state,
        quote: {
          ...state.quote,
          data: action.payload.value.isError ? null : state.quote.data,
        },
        form: {
          ...state.form,
          [action.payload.field]: action.payload.value,
        },
      };
    case actionTypes.QUOTE_ERROR:
      return {
        ...state,
        quote: {
          ...state.quote,
          isError: true,
          isFetching: false,
          data: action.payload,
        },
      };

    case actionTypes.QUOTE_SUCCESS:
      return {
        ...state,
        quote: {
          ...state.quote,
          isError: false,
          isFetching: false,
          data: action.payload,
        },
      };
    case actionTypes.QUOTE_FETCH:
      return {
        ...state,
        quote: {
          ...state.quote,
          isError: false,
          isFetching: true,
        },
      };
    default:
      return state;
  }
}
