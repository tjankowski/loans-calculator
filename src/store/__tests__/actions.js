jest.mock("lodash.debounce", () => (callback) => callback);
jest.mock("../../api");
import { useActions, action } from "../actions";
import quote from "../../api";
import { initialState } from "../reducers";
import { FormFieldsValueRange, FormFields } from "../constants";
import debounce from "lodash.debounce";

it("should create action for given arguments", () => {
  const data = Date.now();
  expect(action("action", data)).toEqual({
    type: "action",
    payload: data,
  });
});

describe("Test quote action:", () => {
  it("should handle quote fetching success", async () => {
    const dispatch = jest.fn();
    quote.mockImplementation(() => Promise.resolve("DATA"));
    const actions = useActions(initialState, dispatch);
    await actions.quote(
      FormFieldsValueRange[FormFields.amount][0],
      FormFieldsValueRange[FormFields.duration][0]
    );
    expect(dispatch.mock.calls.length).toBe(2);
    expect(dispatch.mock.calls[0]).toMatchSnapshot();
    expect(dispatch.mock.calls[1]).toMatchSnapshot();
  });

  it("should handle quote fetching error", async () => {
    const dispatch = jest.fn();
    quote.mockImplementation(() => Promise.reject("ERROR"));
    const actions = useActions(initialState, dispatch);
    await actions.quote(
      FormFieldsValueRange[FormFields.amount][0],
      FormFieldsValueRange[FormFields.duration][0]
    );
    expect(dispatch.mock.calls.length).toBe(2);
    expect(dispatch.mock.calls[0]).toMatchSnapshot();
    expect(dispatch.mock.calls[1]).toMatchSnapshot();
  });
});

describe("Test amount field change:", () => {
  it("should handle action without errors for min value", async () => {
    const dispatch = jest.fn();
    quote.mockImplementation(() => Promise.resolve("DATA"));
    const actions = useActions(initialState, dispatch);
    await actions.onFormChange(
      FormFields.amount,
      FormFieldsValueRange[FormFields.amount][0]
    );
    expect(dispatch.mock.calls.length).toBe(3);
    expect(dispatch.mock.calls[0]).toMatchSnapshot();
    expect(dispatch.mock.calls[1]).toMatchSnapshot();
    expect(dispatch.mock.calls[2]).toMatchSnapshot();
  });

  it("should handle action without errors for max value", async () => {
    const dispatch = jest.fn();
    quote.mockImplementation(() => Promise.resolve("DATA"));
    const actions = useActions(initialState, dispatch);
    await actions.onFormChange(
      FormFields.amount,
      FormFieldsValueRange[FormFields.amount][1]
    );
    expect(dispatch.mock.calls.length).toBe(3);
    expect(dispatch.mock.calls[0]).toMatchSnapshot();
    expect(dispatch.mock.calls[1]).toMatchSnapshot();
    expect(dispatch.mock.calls[2]).toMatchSnapshot();
  });

  it("should handle action with fetch errors for min value", async () => {
    const dispatch = jest.fn();
    quote.mockImplementation(() => Promise.reject("ERROR"));
    const actions = useActions(initialState, dispatch);
    await actions.onFormChange(
      FormFields.amount,
      FormFieldsValueRange[FormFields.amount][0]
    );
    expect(dispatch.mock.calls.length).toBe(3);
    expect(dispatch.mock.calls[0]).toMatchSnapshot();
    expect(dispatch.mock.calls[1]).toMatchSnapshot();
    expect(dispatch.mock.calls[2]).toMatchSnapshot();
  });

  it("should handle action with fetch errors for max value", async () => {
    const dispatch = jest.fn();
    quote.mockImplementation(() => Promise.reject("ERROR"));
    const actions = useActions(initialState, dispatch);
    await actions.onFormChange(
      FormFields.amount,
      FormFieldsValueRange[FormFields.amount][1]
    );
    expect(dispatch.mock.calls.length).toBe(3);
    expect(dispatch.mock.calls[0]).toMatchSnapshot();
    expect(dispatch.mock.calls[1]).toMatchSnapshot();
    expect(dispatch.mock.calls[2]).toMatchSnapshot();
  });

  it("should mark validation error value below min", async () => {
    const dispatch = jest.fn();
    quote.mockImplementation(() => Promise.resolve("DATA"));
    const actions = useActions(initialState, dispatch);
    await actions.onFormChange(
      FormFields.amount,
      FormFieldsValueRange[FormFields.amount][0] - 1
    );
    expect(dispatch.mock.calls.length).toBe(1);
    expect(dispatch.mock.calls[0]).toMatchSnapshot();
  });
  it("should mark validation error value above max", async () => {
    const dispatch = jest.fn();
    quote.mockImplementation(() => Promise.resolve("DATA"));
    const actions = useActions(initialState, dispatch);
    await actions.onFormChange(
      FormFields.amount,
      FormFieldsValueRange[FormFields.amount][1] + 1
    );
    expect(dispatch.mock.calls.length).toBe(1);
    expect(dispatch.mock.calls[0]).toMatchSnapshot();
    expect(dispatch.mock.calls[0][0].payload.value.value).toBe(
      FormFieldsValueRange[FormFields.amount][1]
    );
  });
});

describe("Test duration field change:", () => {
  it("should handle action without errors for min value", async () => {
    const dispatch = jest.fn();
    quote.mockImplementation(() => Promise.resolve("DATA"));
    const actions = useActions(initialState, dispatch);
    await actions.onFormChange(
      FormFields.duration,
      FormFieldsValueRange[FormFields.duration][0]
    );
    expect(dispatch.mock.calls.length).toBe(3);
    expect(dispatch.mock.calls[0]).toMatchSnapshot();
    expect(dispatch.mock.calls[1]).toMatchSnapshot();
    expect(dispatch.mock.calls[2]).toMatchSnapshot();
  });

  it("should handle action without errors for max value", async () => {
    const dispatch = jest.fn();
    quote.mockImplementation(() => Promise.resolve("DATA"));
    const actions = useActions(initialState, dispatch);
    await actions.onFormChange(
      FormFields.duration,
      FormFieldsValueRange[FormFields.duration][1]
    );
    expect(dispatch.mock.calls.length).toBe(3);
    expect(dispatch.mock.calls[0]).toMatchSnapshot();
    expect(dispatch.mock.calls[1]).toMatchSnapshot();
    expect(dispatch.mock.calls[2]).toMatchSnapshot();
  });

  it("should handle action with fetch errors for min value", async () => {
    const dispatch = jest.fn();
    quote.mockImplementation(() => Promise.reject("ERROR"));
    const actions = useActions(initialState, dispatch);
    await actions.onFormChange(
      FormFields.duration,
      FormFieldsValueRange[FormFields.duration][0]
    );
    expect(dispatch.mock.calls.length).toBe(3);
    expect(dispatch.mock.calls[0]).toMatchSnapshot();
    expect(dispatch.mock.calls[1]).toMatchSnapshot();
    expect(dispatch.mock.calls[2]).toMatchSnapshot();
  });

  it("should handle action with fetch errors for max value", async () => {
    const dispatch = jest.fn();
    quote.mockImplementation(() => Promise.reject("ERROR"));
    const actions = useActions(initialState, dispatch);
    await actions.onFormChange(
      FormFields.duration,
      FormFieldsValueRange[FormFields.duration][1]
    );
    expect(dispatch.mock.calls.length).toBe(3);
    expect(dispatch.mock.calls[0]).toMatchSnapshot();
    expect(dispatch.mock.calls[1]).toMatchSnapshot();
    expect(dispatch.mock.calls[2]).toMatchSnapshot();
  });

  it("should mark validation error value below min", async () => {
    const dispatch = jest.fn();
    quote.mockImplementation(() => Promise.resolve("DATA"));
    const actions = useActions(initialState, dispatch);
    await actions.onFormChange(
      FormFields.duration,
      FormFieldsValueRange[FormFields.duration][0] - 1
    );
    expect(dispatch.mock.calls.length).toBe(1);
    expect(dispatch.mock.calls[0]).toMatchSnapshot();
  });
  it("should mark validation error value above max", async () => {
    const dispatch = jest.fn();
    quote.mockImplementation(() => Promise.resolve("DATA"));
    const actions = useActions(initialState, dispatch);
    await actions.onFormChange(
      FormFields.duration,
      FormFieldsValueRange[FormFields.duration][1] + 1
    );
    expect(dispatch.mock.calls.length).toBe(1);
    expect(dispatch.mock.calls[0]).toMatchSnapshot();
    expect(dispatch.mock.calls[0][0].payload.value.value).toBe(
      FormFieldsValueRange[FormFields.duration][1]
    );
  });
});
