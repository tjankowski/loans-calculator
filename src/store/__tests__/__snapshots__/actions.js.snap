// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test amount field change: should handle action with fetch errors for max value 1`] = `
Array [
  Object {
    "payload": Object {
      "field": "amount",
      "value": Object {
        "isError": false,
        "value": 5000,
      },
    },
    "type": "form.input",
  },
]
`;

exports[`Test amount field change: should handle action with fetch errors for max value 2`] = `
Array [
  Object {
    "payload": Object {
      "amount": 5000,
      "duration": 24,
    },
    "type": "quote.fetch",
  },
]
`;

exports[`Test amount field change: should handle action with fetch errors for max value 3`] = `
Array [
  Object {
    "payload": "ERROR",
    "type": "quote.error",
  },
]
`;

exports[`Test amount field change: should handle action with fetch errors for min value 1`] = `
Array [
  Object {
    "payload": Object {
      "field": "amount",
      "value": Object {
        "isError": false,
        "value": 1000,
      },
    },
    "type": "form.input",
  },
]
`;

exports[`Test amount field change: should handle action with fetch errors for min value 2`] = `
Array [
  Object {
    "payload": Object {
      "amount": 1000,
      "duration": 24,
    },
    "type": "quote.fetch",
  },
]
`;

exports[`Test amount field change: should handle action with fetch errors for min value 3`] = `
Array [
  Object {
    "payload": "ERROR",
    "type": "quote.error",
  },
]
`;

exports[`Test amount field change: should handle action without errors for max value 1`] = `
Array [
  Object {
    "payload": Object {
      "field": "amount",
      "value": Object {
        "isError": false,
        "value": 5000,
      },
    },
    "type": "form.input",
  },
]
`;

exports[`Test amount field change: should handle action without errors for max value 2`] = `
Array [
  Object {
    "payload": Object {
      "amount": 5000,
      "duration": 24,
    },
    "type": "quote.fetch",
  },
]
`;

exports[`Test amount field change: should handle action without errors for max value 3`] = `
Array [
  Object {
    "payload": "DATA",
    "type": "quote.success",
  },
]
`;

exports[`Test amount field change: should handle action without errors for min value 1`] = `
Array [
  Object {
    "payload": Object {
      "field": "amount",
      "value": Object {
        "isError": false,
        "value": 1000,
      },
    },
    "type": "form.input",
  },
]
`;

exports[`Test amount field change: should handle action without errors for min value 2`] = `
Array [
  Object {
    "payload": Object {
      "amount": 1000,
      "duration": 24,
    },
    "type": "quote.fetch",
  },
]
`;

exports[`Test amount field change: should handle action without errors for min value 3`] = `
Array [
  Object {
    "payload": "DATA",
    "type": "quote.success",
  },
]
`;

exports[`Test amount field change: should mark validation error value above max 1`] = `
Array [
  Object {
    "payload": Object {
      "field": "amount",
      "value": Object {
        "isError": true,
        "value": 5000,
      },
    },
    "type": "form.input",
  },
]
`;

exports[`Test amount field change: should mark validation error value below min 1`] = `
Array [
  Object {
    "payload": Object {
      "field": "amount",
      "value": Object {
        "isError": true,
        "value": 999,
      },
    },
    "type": "form.input",
  },
]
`;

exports[`Test duration field change: should handle action with fetch errors for max value 1`] = `
Array [
  Object {
    "payload": Object {
      "field": "duration",
      "value": Object {
        "isError": false,
        "value": 36,
      },
    },
    "type": "form.input",
  },
]
`;

exports[`Test duration field change: should handle action with fetch errors for max value 2`] = `
Array [
  Object {
    "payload": Object {
      "amount": 2000,
      "duration": 36,
    },
    "type": "quote.fetch",
  },
]
`;

exports[`Test duration field change: should handle action with fetch errors for max value 3`] = `
Array [
  Object {
    "payload": "ERROR",
    "type": "quote.error",
  },
]
`;

exports[`Test duration field change: should handle action with fetch errors for min value 1`] = `
Array [
  Object {
    "payload": Object {
      "field": "duration",
      "value": Object {
        "isError": false,
        "value": 6,
      },
    },
    "type": "form.input",
  },
]
`;

exports[`Test duration field change: should handle action with fetch errors for min value 2`] = `
Array [
  Object {
    "payload": Object {
      "amount": 2000,
      "duration": 6,
    },
    "type": "quote.fetch",
  },
]
`;

exports[`Test duration field change: should handle action with fetch errors for min value 3`] = `
Array [
  Object {
    "payload": "ERROR",
    "type": "quote.error",
  },
]
`;

exports[`Test duration field change: should handle action without errors for max value 1`] = `
Array [
  Object {
    "payload": Object {
      "field": "duration",
      "value": Object {
        "isError": false,
        "value": 36,
      },
    },
    "type": "form.input",
  },
]
`;

exports[`Test duration field change: should handle action without errors for max value 2`] = `
Array [
  Object {
    "payload": Object {
      "amount": 2000,
      "duration": 36,
    },
    "type": "quote.fetch",
  },
]
`;

exports[`Test duration field change: should handle action without errors for max value 3`] = `
Array [
  Object {
    "payload": "DATA",
    "type": "quote.success",
  },
]
`;

exports[`Test duration field change: should handle action without errors for min value 1`] = `
Array [
  Object {
    "payload": Object {
      "field": "duration",
      "value": Object {
        "isError": false,
        "value": 6,
      },
    },
    "type": "form.input",
  },
]
`;

exports[`Test duration field change: should handle action without errors for min value 2`] = `
Array [
  Object {
    "payload": Object {
      "amount": 2000,
      "duration": 6,
    },
    "type": "quote.fetch",
  },
]
`;

exports[`Test duration field change: should handle action without errors for min value 3`] = `
Array [
  Object {
    "payload": "DATA",
    "type": "quote.success",
  },
]
`;

exports[`Test duration field change: should mark validation error value above max 1`] = `
Array [
  Object {
    "payload": Object {
      "field": "duration",
      "value": Object {
        "isError": true,
        "value": 36,
      },
    },
    "type": "form.input",
  },
]
`;

exports[`Test duration field change: should mark validation error value below min 1`] = `
Array [
  Object {
    "payload": Object {
      "field": "duration",
      "value": Object {
        "isError": true,
        "value": 5,
      },
    },
    "type": "form.input",
  },
]
`;

exports[`Test quote action: should handle quote fetching error 1`] = `
Array [
  Object {
    "payload": Object {
      "amount": 1000,
      "duration": 6,
    },
    "type": "quote.fetch",
  },
]
`;

exports[`Test quote action: should handle quote fetching error 2`] = `
Array [
  Object {
    "payload": "ERROR",
    "type": "quote.error",
  },
]
`;

exports[`Test quote action: should handle quote fetching success 1`] = `
Array [
  Object {
    "payload": Object {
      "amount": 1000,
      "duration": 6,
    },
    "type": "quote.fetch",
  },
]
`;

exports[`Test quote action: should handle quote fetching success 2`] = `
Array [
  Object {
    "payload": "DATA",
    "type": "quote.success",
  },
]
`;
